[full-ci] [tests-only] Max and Default Quotas

kobergj opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the oCIS component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next version of oCIS.

Please set the following labels:

- Set label "Status:Needs-Review" for review or "Status:In-Progress" if the PR still has open tasks.
- Assignment: assign to self
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
We can now define default and maximum quota amounts for personal and project spaces

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- test environment:
- test case 1:
- test case 2:
- ...

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in owncloud.github.io/ -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

kobergj commented:
@mmattel I added some configuration values in this PR. (Only 3). Could you have a look if my descriptions are sufficient?

mmattel commented:
Technically the description text is ok 👍 
Two questions: pure bytes could be hard to read. Additional values to shorten it like G|M|K ?
What is the base. Is it 1000 or 1024?

kobergj commented:
@mmattel Thanks for input! You are absolutely right. I added possibility to use strings such as `10GB` or `100PiB` and also updated the description accordingly.
Now when using decimal abbreviations (`KB`, `MB`, `GB`, ...) base 1000 will be used and when using binary abbreviations (`KiB`, `MiB`, ...) then the base is 1024

Could you check again please?

mmattel commented:
When you sanitize/validate the string, do you check for numbers only + size extension?
Beside the size what does an admin get as error message when you have a string like 1.024GB? (see the dot between the numbers)

phil-davis commented on services/proxy/pkg/middleware/options.go:
```diff
@@ -212,3 +214,10 @@ func JWKSOptions(jo config.JWKS) Option {
 	}
 }
+
+// DefaultSpaceQuota ssets the default space quota for personal spaces
```
```suggestion
// DefaultSpaceQuota sets the default space quota for personal spaces
```

kulmann commented on services/graph/pkg/config/config.go:
```diff
@@ -31,7 +31,9 @@ type Config struct {
 	WebDavBase                      string `yaml:"webdav_base" env:"OCIS_URL;GRAPH_SPACES_WEBDAV_BASE" desc:"The public facing URL of WebDAV."`
 	WebDavPath                      string `yaml:"webdav_path" env:"GRAPH_SPACES_WEBDAV_PATH" desc:"The WebDAV subpath for spaces."`
-	DefaultQuota                    string `yaml:"default_quota" env:"GRAPH_SPACES_DEFAULT_QUOTA" desc:"The default quota in bytes."`
+	DefaultQuota                    string `yaml:"default_quota" env:"GRAPH_SPACES_DEFAULT_QUOTA" desc:"The default quota for project spaces in bytes. Also understands common abbreviations such as '100KB', '300KiB', '1024MiB', '4GB', ..."`
+	MaxQuotaPersonal                string `yaml:"max_quota_personal" env:"GRAPH_SPACES_MAX_QUOTA_PERSONAL" desc:"The maximum quota for a personal space in bytes. Also understands common abbreviations such as '100KB', '300KiB', '1024MiB', '4GB', ..."`
+	MaxQuotaProject                 string `yaml:"max_quota_project" env:"GRAPH_SPACES_MAX_QUOTA_PROJECT" desc:"The maximum quota for a project space in bytes. Also understands common abbreviations such as '100KB', '300KiB', '1024MiB', '4GB', ..."`
```
I liked the example string. Full listing of the abbreviations makes sense. But can we add one example additionally, so that admins can read the `syntax` by example? (I know, it's not hard, but an example makes me feel safer as a consumer of this).

mmattel replied:
Added

kulmann commented:
What's the plan for the API, keep it in bytes or also expect the string format that you apply to the config now?

kobergj commented:
@mmattel There is no sanitization on the string done. It expects to see a number followed by a known abbreviation. If it can't parse the values it will error level log like this: https://github.com/owncloud/ocis/pull/4354/files#diff-fedfe1baf57a1b40bed18423f6f4d1ee0ed8ebf79ddcda8e94c6ec2decd5b210R70

In your case the error would be `unknown unit '.024GB'. Use common abbreviations such as KB, MiB, GB` which is the correct error message from the service point of view.

And yes a quota of 0 means unlimited. I added that to the description.

@kulmann For now I didn't change the API, but since the package is backwards compatible (it assumes bytes if no unit is given) we could easily add that to all endpoints. Would it make web logic simpler if we would do so?

fschade reviewed (approved):
🚀
